components/core/src/ffi/ir_stream/decoding_methods.hpp

```diff
@@ -31,6 +48,58 @@ namespace ffi::ir_stream {
      */
     IRErrorCode get_encoding_type (ReaderInterface& ir_buf, bool& is_four_bytes_encoding);
 
+    /**
+     * Parse logtypes, dictionary variables and encoded variables
+     * from the next encoded IR message. Returns the parsed tokens by
+     * reference
+     * @tparam encoded_variable_t
+     * @param reader
+     * @param logtype
+     * @param encoded_vars
+     * @param dict_vars
+     * @param timestamp
+     * @return IRErrorCode_Success on success
+     * @return IRErrorCode_Corrupted_IR if reader contains invalid IR
+     * @return IRErrorCode_Incomplete_IR if reader doesn't contain enough data
+     */
+    template <typename encoded_variable_t>
+    IRErrorCode generic_parse_tokens (ReaderInterface& reader,
+                                      std::string& logtype,
+                                      std::vector<encoded_variable_t>& encoded_vars,
+                                      std::vector<std::string>& dict_vars,
+                                      epoch_time_ms_t& timestamp);
+
+    /**
+     * Decodes the message consists of the tokens and calls the given methods
+     * to handle specific components of the message.
+     * @tparam encoded_variable_t Type of the encoded variable
+     * @tparam ConstantHandler Method to handle constants. Signature:
+     * (const std::string&, size_t, size_t) -> void
+     * @tparam EncodedIntHandler Method to handle encoded integers.
+     * Signature: (encoded_variable_t) -> void
+     * @tparam EncodedFloatHandler Method to handle encoded float.
+     * Signature: (encoded_variable_t) -> void
+     * @tparam DictVarHandler Method to handle dictionary variables.
+     * Signature: (const std::string&) -> void
+     * @param logtype
+     * @param encoded_vars
+     * @param dict_vars
+     * @param constant_handler
+     * @param encoded_int_handler
+     * @param encoded_float_handler
+     * @param dict_var_handler
+     * @throw DecodingException if the message can not be decoded properly
+     */
+    template <typename encoded_variable_t, typename ConstantHandler,
+            typename EncodedIntHandler, typename EncodedFloatHandler, typename DictVarHandler>
+    void generic_decode_message (const std::string& logtype,
```

Comment:
Personally, I feel the function name and description are a little bit confusing. The entire decoding process consists of two parts: parsing the encoded buffer and recovering the original message. This function only does the recovery (or true... decoding).
Suggestions:
1. Structure logtype, encoded_vars, and dict_vars into one class and name it as  sth like IRMessage
2. generic_parse_tokens and generic_decode_message can be two member functions of this class.

Reply:
Yeah the function name might need to be repharsed. maybe "generic_decode_tokens" or something similar.

A class would help. I have something similar (yet a little more sophsticated) in ParsedIRMessage, was not sure whether I should use it in the decoding_method. Kirk, just want to know if you see any concern

For 2. if there weren't those lambda's, I would agree with you. But now the code definitely does more than restricting its functionality to IRMessage so I am not sure if I would agree.

Reply:
After discussing offline, we agreed that makeing a new class IRMessage makes sense but it should not own generic_parse_tokens and generic_decode_message functions
Rather, it should act as a struct-like class and be an argument of those two functions